### PR TITLE
refactor(popupmenu.c): misc

### PR DESCRIPTION
**Pull request by glepnir**
replace some if blocks and combine user attr abstract to inline function, If need to add user_menu in the future, there don't need to write duplicate code, common expressions use variables to calculate in advance, position calculations might be more readable ...

### CODE REVIEW

**Comment by chrisbra** [comment]
thanks

File: src/popupmenu.c [hunk 1]
```diff
@@ -911,6 +917,7 @@ pum_set_selected(int n, int repeat UNUSED)
 {
     int	    resized = FALSE;
     int	    context = pum_height / 2;
+    int	    scroll_offset = pum_selected - pum_height;
```
Comment (zeertzjq) [inline src/popupmenu.c:920]: Shouldn't this come after the `pum_selected = n`?

File: src/popupmenu.c [hunk 2]
```diff
@@ -938,37 +945,24 @@ pum_set_selected(int n, int repeat UNUSED)
...
-	    {
-		pum_first += pum_height - 2;
-		if (pum_first < pum_selected - pum_height + 1)
-		    pum_first = pum_selected - pum_height + 1;
-	    }
+	    if (pum_first < scroll_offset + 3)
+		pum_first = MAX(pum_first, scroll_offset + 1);
 	    else
-		pum_first = pum_selected - pum_height + 1;
+		pum_first = scroll_offset + 1;
```
Comment (zeertzjq) [inline src/popupmenu.c:955]: Where did the `pum_first += pum_height - 2` go?

Comment (glepnir) [reply src/popupmenu.c:955]: fix in https://github.com/vim/vim/pull/16283

**Pull request closed without merging**